File: dsa_leetcode_beginner/0268-missing-number/0268-missing-number.java

public class MissingNumber {
    public static int missingNumber(int[] nums) {
        int xor = 0;
        int n = nums.length;

        // XOR all indices and all elements in nums
        for (int i = 0; i < n; i++) {
            xor = xor ^ i ^ nums[i];
        }

        // Finally XOR with n (since the range is from 0 to n)
        xor = xor ^ n;

        return xor;
    }

    public static void main(String[] args) {
        int[] nums = {3, 0, 1};
        System.out.println(missingNumber(nums));  // Output: 2
    }
}

File: dsa_leetcode_beginner/0268-missing-number/0268-missing-number.cpp

#include <iostream>
#include <vector>

using namespace std;

int missingNumber(vector<int>& nums) {
    int xor_result = 0;
    int n = nums.size();

    // XOR all indices and all elements in nums
    for (int i = 0; i < n; i++) {
        xor_result = xor_result ^ i ^ nums[i];
    }

    // Finally XOR with n (since the range is from 0 to n)
    xor_result = xor_result ^ n;

    return xor_result;
}

int main() {
    vector<int> nums = {3, 0, 1};
    cout << missingNumber(nums) << endl;  // Output: 2

    return 0;
}
